README.md
# PéK-Next

**The administration system for Schönherz Student Hostel and [SVIE](http://svie.hu/)**

[![Build Status](https://travis-ci.org/kir-dev/pek-next.png?branch=master)](https://travis-ci.org/kir-dev/pek-next)
[![Code Climate](https://codeclimate.com/github/kir-dev/pek-next.png)](https://codeclimate.com/github/kir-dev/pek-next)
[![Dependency Status](https://gemnasium.com/kir-dev/pek-next.png)](https://gemnasium.com/kir-dev/pek-next)

## Requirements

- Ruby 2.5.7
- Postgresql 9.6
- Node (asset compiling)
- Redis (optional)

or

- Docker

## Installing requirements

### Debian derivatives

###### Packages

```bash
# Redis is optional
sudo apt install postgresql-9.6 libpq-dev nodejs redis-server
```

###### Ruby 2.5.7

Use [asdf](https://asdf-vm.github.io/asdf/#/core-manage-asdf-vm) with [ruby plugin](https://github.com/asdf-vm/asdf-ruby) or [rbenv](https://github.com/rbenv/rbenv). Install Ruby 2.5.7 and set executable version. You can check current ruby version with `ruby -v`

### MacOS

###### Brew

Easiest way to install [brew](https://brew.sh), then install required packages.

```bash
# Redis is optional
brew install asdf postgresql@9.6 redis
brew services start postgresql@9.6
brew services start redis
```

###### Ruby 2.5.7 and Node

Add [ruby](https://github.com/asdf-vm/asdf-ruby) and [node](https://github.com/asdf-vm/asdf-nodejs) plugin to asdf. Install Ruby 2.5.7 and set executable version. You can check current ruby version with `ruby -v`. Install node too.

## Setting up

###### The source code and dependencies

```bash
git clone https://github.com/kir-dev/pek-next.git
cd pek-next
gem install bundler
bundle install
```

###### Environment

Create a `.env` file using `.env.example` and replace the values with real ones.

###### The database

```bash
sudo su postgres
psql -c 'create user "pek-next" with superuser password '\''pek-next'\'';'
```

###### Init database

_As your own user_

```bash
rake db:setup
```

## Running

###### The server

```bash
rails s
```

###### Worker (optional, requires redis)

```bash
bundle exec sidekiq
```

###### The tests (of course)

```bash
rake test
```

## Deployment

Easiest way for deployment is docker-compose.

Copy `.env` from `.env.example`, add `APP_ID` and `APP_SECRET` according to auth.sch and generate a `SECRET_KEY_BASE` using `bundle exec rake secret`.

Then run the following commands:

```bash
# These volumes are not necessary and could be removed from docker-compose, but a named volume easier to find later on
docker volume create pek_public
docker volume create pek_database
docker-compose up --build
```

After creating, while the containers are running run the following commands:

```bash
# This is only necessary at new setups
docker-compose run web bash -c "bundle exec rake db:setup"

# This is only necessary after pending migrations
docker-compose run web bash -c "bundle exec rake db:migrate"

# This is required at new setups and after changing in assets
docker-compose run web bash -c "bundle exec rake assets:precompile"
```

## Problems you may encounter and the solutions

###### Ruby cannot build the native extensions

```bash
sudo apt install ruby-dev
```

###### Rbenv install only gives ruby-build usage instructions

```bash
rbenv install -v 2.5.7
```

###### Rbenv doesn't modify your ruby version

```bash
echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.bashrc
echo 'eval "$(rbenv init -)"' >> ~/.bashrc
source ~/.bashrc
```

###### Rails command is not recognized after install

_Restart your terminal_

by [Kir-Dev Team](http://kir-dev.sch.bme.hu/)

### Special thanks for

[![Rollbar](public/img/rollbar.png)](https://rollbar.com/)


CHANGELOG.md
# Changelog

## 2.0.0
- Rails 5.0 🎉
- Add group types enum (`:group`, `:team`, `:resort`)
- Add Hungarian README
- Under the hood
  - Shrink Docker image
  - Bump Ruby version to 2.5.7
  - Add `simplecov`

## v1.9.1
- Fix error on showing history for groups without any evaluation
- Under the hood
  - Upgrade Rails patch version
  - Fix failing migration task for new member posts
  - Init RSpec

## v1.9
- Unify new member post for groups
- Fix link for SVIE admin page
- Fix displaying point details for judgement page
- Increase max responsibility point to 50
- Under the hood
  - Start changing fixtures to FactoryBot

## v1.8
- Fix link duplications on group history
- Add SVIE hierarchy view
- Fix indicating new groups as 'probably inactive'
- Under the hood
  - Change unique constraint on screen name to case-insensitive

## v1.7
- Add comments on point table
- Redesign point table on judgement page
  - Update style from evaluation table
  - Add point sums & averages
  - Add sorting
- Add support help text, remove popup
- Update README
- Fix sanitizing on profile page
- Change membership handling
  - Update buttons after successful action
  - Add confirmation before archiving user
- Add custom error pages
- Update SVIE related stuff
  - Async requests on admin page
  - Move admin related stuff to module
- Under the hood
  - Refactor link decorations
  - Update robots.txt
  - Optimize queries on judgement page
  - Add docker support
  - Refactor associated models delete policy

## v1.6.1
- Update evaluation table sum
- Fix missing database and model validations

## v1.6
- Add UIKit date picker for date inputs
- Change 401 (unauthorized) to 403 (forbidden)
- **Add delegates counting and count editing**
- Improve evaluation related pages
  - Add user hiding
  - Make user column sticky
  - Add arrow navigation in table
  - Optimize for smaller/mobile screens
- Under the hood
  - refactor codebase
  - optimize evaluation related queries, decrease loading time ~66%

## v1.5
- Add link to user's mail and phone number, thanks @kissadamfkut
- Add redirect for old style VIR id links
- Add group history
- Change room number to string
- Fix login date update
- Under the hood:
  - Remove ES6 syntax to avoid failing asset building
  - Refactor membership related stuff
  - Fix seeder

## v1.4
- Adjust uploaded picture quality
- Add view settings for list/grid, pagination and show pictures
- Updated styles
- Fix null error in point details
- Add point history calculation
- Under the hood:
  - Refactor semester class

## v1.3.1
- Fix point details sum on judgement page
- Add evaluation messaging for group leader
- Fix missing date of birth field and validation on SVIE application
- Fix evaluation editing
- Add group parent info to group page

## v1.3
- Show evaluations for RVT members includin point request, entry requests, principles and connecting messages
- Add ability to create evaluation judgement
- Add share button for profile page
- Add redirecting to opened url after login
- Add more clear description for group explanation for evaluation
- Add ability to modify evaluation after deny
- Add ability for submitting evaluation and remove auto submission logic
- Fix evaluation's group explanation to justification
- Add validations for point details (between 0 and principle's max)
- Update point request sum (20/30 rule, rounding)

## v1.2.1
- Fix entry request justification editing, add redirect back, when no entry request given
- Fix point table, all group member can get point, not only SVIE members
- Fix entry request's justification, remove them when type is 'KDO' at season change
- Fix principle editing, add description input
- Add principle summary popup to evaluation table

## v1.2
- Add creation for evaluation description
- Add creation for principles
- Add creation for evaluation
- Add creation for justification
- Fix SVIE contraction text
- Fix archived member listing
- Add info for group leaders about archiving
- Remove archived members from authsch
- Add fallback location to redirect back, when no referrer provided
- Fix user point history nil error

## v1.1
- Fix group membership check
- Add separated listing for active groups
- Rename SVIE member types to: 'Rendes tag', 'Külső tag', 'Öreg tag'
- Fix contradiction between SVIE application and primary group select
- Update registration page, add explanation for screen_name
- Fix screen name validation, disallow slashes
- Fix new post type creation, post types belong to groups, won't be globals
- Add group name to user's post editing page
- Add validation for post types, restrict empty name
- Fix broken cropped photo path


db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20161124163851) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_table "belepoigenyles", id: :bigserial, force: :cascade do |t|
    t.string  "belepo_tipus",       limit: 255
    t.text    "szoveges_ertekeles"
    t.integer "ertekeles_id",       limit: 8,   null: false
    t.integer "usr_id",             limit: 8
  end

  add_index "belepoigenyles", ["belepo_tipus"], name: "bel_tipus_idx", using: :btree

  create_table "ertekeles_uzenet", id: :bigserial, force: :cascade do |t|
    t.datetime "feladas_ido"
    t.text     "uzenet"
    t.integer  "felado_usr_id", limit: 8
    t.integer  "group_id",      limit: 8
    t.string   "semester",      limit: 9,                 null: false
    t.boolean  "from_system",             default: false
  end

  add_index "ertekeles_uzenet", ["felado_usr_id"], name: "fki_felado_usr_id", using: :btree
  add_index "ertekeles_uzenet", ["group_id"], name: "fki_group_id", using: :btree

  create_table "ertekelesek", id: :bigserial, force: :cascade do |t|
    t.string   "belepoigeny_statusz", limit: 255
    t.datetime "feladas"
    t.string   "pontigeny_statusz",   limit: 255
    t.string   "semester",            limit: 9,                   null: false
    t.text     "szoveges_ertekeles",                              null: false
    t.datetime "utolso_elbiralas"
    t.datetime "utolso_modositas"
    t.integer  "elbiralo_usr_id",     limit: 8
    t.integer  "grp_id",              limit: 8
    t.integer  "felado_usr_id",       limit: 8
    t.text     "pontozasi_elvek",                 default: "",    null: false
    t.integer  "next_version",        limit: 8
    t.text     "explanation"
    t.integer  "optlock",                         default: 0,     null: false
    t.boolean  "is_considered",                   default: false, null: false
  end

  add_index "ertekelesek", ["grp_id", "semester", "next_version"], name: "unique_idx", unique: true, using: :btree
  add_index "ertekelesek", ["next_version"], name: "next_version_idx", using: :btree
  add_index "ertekelesek", ["semester"], name: "ert_semester_idx", using: :btree

  create_table "groups", primary_key: "grp_id", force: :cascade do |t|
    t.text    "grp_name",                                        null: false
    t.string  "grp_type",             limit: 20,                 null: false
    t.integer "grp_parent",           limit: 8
    t.string  "grp_state",                       default: "akt"
    t.text    "grp_description"
    t.string  "grp_webpage",          limit: 64
    t.string  "grp_maillist",         limit: 64
    t.string  "grp_head",             limit: 48
    t.integer "grp_founded"
    t.boolean "grp_issvie",                      default: false, null: false
    t.integer "grp_svie_delegate_nr"
    t.boolean "grp_users_can_apply",             default: true,  null: false
  end

  add_index "groups", ["grp_id"], name: "groups_grp_id_idx", unique: true, using: :btree
  add_index "groups", ["grp_name"], name: "idx_groups_grp_name", using: :btree
  add_index "groups", ["grp_type"], name: "idx_groups_grp_type", using: :btree

  create_table "grp_membership", id: :bigserial, force: :cascade do |t|
    t.integer "grp_id",           limit: 8
    t.integer "usr_id",           limit: 8
    t.date    "membership_start",           default: "now()"
    t.date    "membership_end"
  end

  add_index "grp_membership", ["grp_id", "usr_id"], name: "unique_memberships", unique: true, using: :btree
  add_index "grp_membership", ["usr_id"], name: "membership_usr_fk_idx", using: :btree

  create_table "im_accounts", id: :bigserial, force: :cascade do |t|
    t.string  "protocol",     limit: 50,  null: false
    t.string  "account_name", limit: 255, null: false
    t.integer "usr_id",       limit: 8
  end

  create_table "log", id: :bigserial, force: :cascade do |t|
    t.integer "grp_id",   limit: 8
    t.integer "usr_id",   limit: 8,                    null: false
    t.date    "evt_date",            default: "now()"
    t.string  "event",    limit: 30,                   null: false
  end

  create_table "lostpw_tokens", primary_key: "usr_id", force: :cascade do |t|
    t.datetime "created"
    t.string   "token",   limit: 64
  end

  add_index "lostpw_tokens", ["token"], name: "lostpw_tokens_token_key", unique: true, using: :btree

  create_table "neptun_list", primary_key: "neptun", force: :cascade do |t|
    t.string  "nev",          limit: 128,                 null: false
    t.date    "szuldat",                                  null: false
    t.string  "education_id", limit: 11
    t.boolean "newbie",                   default: false
  end

  create_table "point_history", id: :bigserial, force: :cascade do |t|
    t.integer "usr_id",   limit: 8, null: false
    t.integer "point",              null: false
    t.string  "semester", limit: 9, null: false
  end

  create_table "pontigenyles", id: :bigserial, force: :cascade do |t|
    t.integer "pont"
    t.integer "ertekeles_id", limit: 8, null: false
    t.integer "usr_id",       limit: 8
  end

  create_table "poszt", id: :bigserial, force: :cascade do |t|
    t.integer "grp_member_id", limit: 8
    t.integer "pttip_id",      limit: 8
  end

  add_index "poszt", ["grp_member_id"], name: "poszt_fk_idx", using: :btree

  create_table "poszttipus", primary_key: "pttip_id", force: :cascade do |t|
    t.integer "grp_id",         limit: 8
    t.string  "pttip_name",     limit: 30,                 null: false
    t.boolean "delegated_post",            default: false
  end

  create_table "spot_images", id: false, force: :cascade do |t|
    t.string "usr_neptun",             null: false
    t.string "image_path", limit: 255, null: false
  end

  add_index "spot_images", ["usr_neptun"], name: "spot_images_usr_neptun_key", unique: true, using: :btree

  create_table "system_attrs", primary_key: "attributeid", force: :cascade do |t|
    t.string "attributename",  limit: 255, null: false
    t.string "attributevalue", limit: 255, null: false
  end

  create_table "temp_belepo", id: false, force: :cascade do |t|
    t.text   "usr_lastname"
    t.text   "usr_firstname"
    t.text   "usr_nickname"
    t.text   "grp_name"
    t.string "belepo_tipus",       limit: 255
    t.text   "szoveges_ertekeles"
  end

  create_table "users", primary_key: "usr_id", force: :cascade do |t|
    t.string   "usr_email",                   limit: 64
    t.string   "usr_neptun"
    t.text     "usr_firstname",                                                    null: false
    t.text     "usr_lastname",                                                     null: false
    t.text     "usr_nickname"
    t.string   "usr_svie_state",              limit: 255, default: "NEMTAG",       null: false
    t.string   "usr_svie_member_type",        limit: 255, default: "NEMTAG",       null: false
    t.integer  "usr_svie_primary_membership", limit: 8
    t.boolean  "usr_delegated",                           default: false,          null: false
    t.boolean  "usr_show_recommended_photo",              default: false,          null: false
    t.string   "usr_screen_name",             limit: 50,                           null: false
    t.date     "usr_date_of_birth"
    t.string   "usr_gender",                  limit: 50,  default: "NOTSPECIFIED", null: false
    t.string   "usr_student_status",          limit: 50,  default: "UNKNOWN",      null: false
    t.string   "usr_mother_name",             limit: 100
    t.string   "usr_photo_path",              limit: 255
    t.string   "usr_webpage",                 limit: 255
    t.string   "usr_cell_phone",              limit: 50
    t.string   "usr_home_address",            limit: 255
    t.string   "usr_est_grad",                limit: 10
    t.string   "usr_dormitory",               limit: 50
    t.string   "usr_room",                    limit: 10
    t.string   "usr_status",                  limit: 8,   default: "INACTIVE",     null: false
    t.string   "usr_password",                limit: 28
    t.string   "usr_salt",                    limit: 12
    t.datetime "usr_lastlogin"
    t.string   "usr_auth_sch_id"
    t.string   "usr_bme_id"
    t.datetime "usr_created_at"
  end

  add_index "users", ["usr_auth_sch_id"], name: "users_usr_auth_sch_id_key", unique: true, using: :btree
  add_index "users", ["usr_bme_id"], name: "users_usr_bme_id_key", unique: true, using: :btree
  add_index "users", ["usr_id"], name: "users_usr_id_idx", unique: true, using: :btree
  add_index "users", ["usr_neptun"], name: "users_usr_neptun_key", unique: true, using: :btree
  add_index "users", ["usr_screen_name"], name: "users_usr_screen_name_key", unique: true, using: :btree

  create_table "usr_private_attrs", id: :bigserial, force: :cascade do |t|
    t.integer "usr_id",    limit: 8,                  null: false
    t.string  "attr_name", limit: 64,                 null: false
    t.boolean "visible",              default: false, null: false
  end

  add_foreign_key "belepoigenyles", "ertekelesek", column: "ertekeles_id", name: "fk_ertekeles_id", on_delete: :cascade
  add_foreign_key "belepoigenyles", "users", column: "usr_id", primary_key: "usr_id", name: "fk4e301ac36958e716"
  add_foreign_key "ertekeles_uzenet", "groups", primary_key: "grp_id", name: "fk_group_id", on_delete: :cascade
  add_foreign_key "ertekeles_uzenet", "users", column: "felado_usr_id", primary_key: "usr_id", name: "fk_felado_usr_id", on_delete: :nullify
  add_foreign_key "ertekelesek", "ertekelesek", column: "next_version", name: "fk_next_version", on_delete: :nullify
  add_foreign_key "ertekelesek", "groups", column: "grp_id", primary_key: "grp_id", name: "fk807db18879696582"
  add_foreign_key "ertekelesek", "users", column: "elbiralo_usr_id", primary_key: "usr_id", name: "fk807db188b31cf015"
  add_foreign_key "ertekelesek", "users", column: "felado_usr_id", primary_key: "usr_id", name: "fk807db18871c0d156"
  add_foreign_key "groups", "groups", column: "grp_parent", primary_key: "grp_id", name: "$1", on_update: :cascade, on_delete: :nullify
  add_foreign_key "grp_membership", "groups", column: "grp_id", primary_key: "grp_id", name: "grp_membership_grp_id_fkey", on_update: :cascade, on_delete: :cascade
  add_foreign_key "grp_membership", "users", column: "usr_id", primary_key: "usr_id", name: "grp_membership_usr_id_fkey", on_update: :cascade, on_delete: :cascade
  add_foreign_key "im_accounts", "users", column: "usr_id", primary_key: "usr_id", name: "im_accounts_usr_id_fkey"
  add_foreign_key "log", "groups", column: "grp_id", primary_key: "grp_id", name: "log_group", on_update: :cascade, on_delete: :cascade
  add_foreign_key "log", "users", column: "usr_id", primary_key: "usr_id", name: "log_user", on_update: :cascade, on_delete: :cascade
  add_foreign_key "lostpw_tokens", "users", column: "usr_id", primary_key: "usr_id", name: "fk1e9df02e5854b081"
  add_foreign_key "point_history", "users", column: "usr_id", primary_key: "usr_id", name: "point_history_usr_id_fkey"
  add_foreign_key "pontigenyles", "ertekelesek", column: "ertekeles_id", name: "fk_ertekeles_id", on_delete: :cascade
  add_foreign_key "pontigenyles", "users", column: "usr_id", primary_key: "usr_id", name: "fkaa1034cd6958e716"
  add_foreign_key "poszt", "grp_membership", column: "grp_member_id", name: "poszt_grp_member_fk", on_update: :cascade, on_delete: :cascade
  add_foreign_key "poszt", "poszttipus", column: "pttip_id", primary_key: "pttip_id", name: "poszt_pttip_fk", on_update: :cascade, on_delete: :cascade
  add_foreign_key "poszttipus", "groups", column: "grp_id", primary_key: "grp_id", name: "poszttipus_opc_csoport", on_update: :cascade, on_delete: :cascade
  add_foreign_key "users", "grp_membership", column: "usr_svie_primary_membership", name: "users_main_group_fkey"
  add_foreign_key "usr_private_attrs", "users", column: "usr_id", primary_key: "usr_id", name: "usr_private_attrs_usr_id_fkey"
end
